src/main/java/org/commcare/session/CommCareSession.java

```diff
@@ -374,6 +374,15 @@ private boolean shouldPopNext(EvaluationContext evalContext) {
         String neededData = getNeededData(evalContext);
         String poppedType = popped == null ? "" : popped.getType();
 
+        if (popped != null) {
+            Boolean wasLastStepCompound = (Boolean) popped.getExtra(SessionFrame.COMPOUND_STEP);
+            Boolean isCurrentStepCompound = (Boolean) frame.getTopStepExtra(SessionFrame.COMPOUND_STEP);
+            Boolean yes = true;
+            if (yes.equals(wasLastStepCompound) && yes.equals(isCurrentStepCompound)) {
```

Comment:
```suggestion
            if (wasLastStepCompound && isCurrentStepCompound) {
```

Reply:
Just realised why I had to use `yes.equals`. Will remove the last commit. 
`wasLastStepCompound ` and `isCurrentStepCompound` can be null so if I do `if (null  && null)` it'll cause NPE.

Reply:
there is not much value in defaulting a boolean to `null` , we should default it to `false` if possible.

Reply:
It come from here: 
https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/suite/model/StackFrameStep.java#L123

and to get a default `false` we'll have to modify that to something like
```
if (SessionFrame.COMPOUND_STEP.equals(key)) {
  Boolean obj = (Boolean) extras.get(key);
  return obj == null ? false : obj
} else {
  return extras.get(key);
}
```

or we can have add another method there: 
```
public boolean isCompoundStep() {
  Boolean obj = (Boolean) extras.get(SessionFrame.COMPOUND_STEP);
  return obj == null ? false : obj
}
```
I like the second option better